pkg/controllers/disruption/orchestration/queue.go

```diff
@@ -0,0 +1,309 @@
+	error
+}
+
+func NewUnrecoverableError(err error) *UnrecoverableError {
+	return &UnrecoverableError{error: err}
+}
+
+func IsUnrecoverableError(err error) bool {
+	if err == nil {
+		return false
+	}
+	var unrecoverableError *UnrecoverableError
+	return errors.As(err, &unrecoverableError)
+}
+
+type Queue struct {
+	workqueue.RateLimitingInterface
+	// providerID -> command, maps a candidate to its command. Each command has a list of candidates that can be used
+	// to map to itself.
+	ProviderIDToCommand map[string]*Command
+
+	kubeClient  client.Client
+	recorder    events.Recorder
+	cluster     *state.Cluster
+	clock       clock.Clock
+	provisioner *provisioning.Provisioner
+}
+
+// NewQueue creates a queue that will asynchronously orchestrate disruption commands
+func NewQueue(kubeClient client.Client, recorder events.Recorder, cluster *state.Cluster, clock clock.Clock,
+	provisioner *provisioning.Provisioner) *Queue {
+	queue := &Queue{
+		RateLimitingInterface: workqueue.NewRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(queueBaseDelay, queueMaxDelay)),
+		ProviderIDToCommand:   map[string]*Command{},
+		kubeClient:            kubeClient,
+		recorder:              recorder,
+		cluster:               cluster,
+		clock:                 clock,
+		provisioner:           provisioner,
+	}
+	return queue
+}
+
+// NewCommand creates a command key and adds in initial data for the orchestration queue.
+func NewCommand(replacements []nodeclaim.Key, candidates []*state.StateNode, timeAdded time.Time, reason string, consolidationType string) *Command {
+	return &Command{
+		ReplacementKeys: lo.Map(replacements, func(key nodeclaim.Key, _ int) *NodeClaimKey {
+			return &NodeClaimKey{Key: key}
+		}),
+		Candidates:        candidates,
+		Method:            reason,
+		ConsolidationType: consolidationType,
+		TimeAdded:         timeAdded,
+	}
+}
+
+func (q *Queue) Name() string {
+	return "disruption.queue"
+}
+
+func (q *Queue) Builder(_ context.Context, m manager.Manager) controller.Builder {
+	return controller.NewSingletonManagedBy(m)
+}
+
+func (q *Queue) Reconcile(ctx context.Context, _ reconcile.Request) (reconcile.Result, error) {
+	// Check if the queue is empty. client-go recommends not using this function to gate the subsequent
+	// get call, but since we're popping items off the queue synchronously, there should be no synchonization
+	// issues.
+	for q.Len() == 0 {
+		return reconcile.Result{RequeueAfter: 1 * time.Second}, nil
+	}
+	// Get command from queue. This waits until queue is non-empty.
+	item, shutdown := q.RateLimitingInterface.Get()
+	if shutdown {
+		return reconcile.Result{}, fmt.Errorf("disruption queue has shut down")
+	}
+	cmd := item.(*Command)
+	defer q.RateLimitingInterface.Done(cmd)
+
+	if err := q.Process(ctx, cmd); err != nil {
+		// If recoverable, re-queue and try again.
+		if !IsUnrecoverableError(err) {
+			// log the error that is causing us to fail so we can bubble it up later if this times out.
```

Comment:
```suggestion
			// store the error that is causing us to fail so we can bubble it up later if this times out.
```

We're not logging it here, right?